pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -839,7 +839,7 @@ protected boolean canEnqueueMessage(Message<T> message) {
 
     protected boolean enqueueMessageAndCheckBatchReceive(Message<T> message) {
         int messageSize = message.size();
-        if (canEnqueueMessage(message) && incomingMessages.offer(message)) {
+        if (isValidConsumerEpoch(message) && canEnqueueMessage(message) && incomingMessages.offer(message)) {
```

Comment:
> If the consumer has pending receive requests. The message will be used to complete the pending receive requests directly.

For users, when having pending receive, It does not call the `redeliverUnacknowledgedMessages` method.  It is only recommended to use the following methods.

``` java
consumer.receive(timeOut);
// do something.
consumer.redeliverUnacknowledgedMessages();
```

Reply:
> And do we need to increase the available permits? Otherwise, the broker might stop dispatching messages to the consumer due to the consumer doesn't have available permits (the messages are skipped)

Yes, we need to increase the available permits. I changed it. and add `testRedeliveryAddEpochAndPermits` unit cover it. PTAL.

Reply:
> For users, when having pending receive, It does not call the redeliverUnacknowledgedMessages method. The following usage is recommended.

@shibd I see. Thanks for the explanation.

It looks like we should not have pending receive requests during the message redelivery. From the understandability and sustainability, we should make it clear that all the messages with an unexpected epoch number will be filtered out. Now, it looks like we must know call `consumer.redeliverUnacknowledgedMessages()` doesn't make any sense before the `receive()` method. If we just read the code, I think a lot of people would think this might be a place to be missed, it is not easy to understand how correctness is guaranteed.

Reply:
Yes, agree. I move the judge epoch to `ConsumerImpl#executeNotifyCallback` method. PATL.